File: CS/1008_Construct_Binary_Search_Tree_from_Preorder_Traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    int[] arr;
    public TreeNode BstFromPreorder(int[] preorder) {
        arr = preorder;
        TreeNode root = new TreeNode();
        
        root = helper(0, preorder.Length);
        
        return root;
    }
    private TreeNode helper(int l, int r){
        if(r - l <= 0){
            return null;
        }
        TreeNode root = new TreeNode();
        root.val = arr[l];
        int mid = l;
        for(int i = l+1; i < r; i++){
            if(arr[i] < root.val){
                mid = i;
            }
            else{
                break;
            }
        }
        
        root.left = helper(l+1, mid+1);
        root.right = helper(mid+1, r);
        
        return root;
    }
}

File: Cpp/1008_Construct_Binary_Search_Tree_from_Preorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        if(preorder.size() > 0){
            return helper(preorder, 0, preorder.size()-1);
        }
        return nullptr;
    }
private:
    TreeNode* helper(vector<int>& preorder, int start, int end){
        if(start <= end){
            TreeNode* result = new TreeNode(preorder[start]);
            int mid = end;
            for(int i = start+1; i < preorder.size(); i++){
                if(preorder[i] > preorder[start]){
                    mid = i-1;
                    break;
                }
            }
            result->left = helper(preorder, start+1, mid);
            result->right = helper(preorder, mid+1, end);
            return result;
        }
        return nullptr;
    }
};
